### PR TITLE
Do not suggest borrow that is already there in fully-qualified call

### LINKED ISSUE #132041: Unhelpful borrowing suggested when wrongly constructing an unsized type
### Code
```rust
fn main() {
    let _x = &str::from("value");
}
```

### Current output
```
error[E0277]: the trait bound `str: From<_>` is not satisfied
 --> src/main.rs:2:15
  |
2 |     let _x = &str::from("value");
  |               ^^^ the trait `From<_>` is not implemented for `str`
  |
help: consider borrowing here
  |
2 |     let _x = &&str::from("value");
  |               +
2 |     let _x = &&mut str::from("value");
  |               ++++

error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> src/main.rs:2:15
  |
2 |     let _x = &str::from("value");
  |               ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `str`
  = note: the return type of a function must have a statically known size

For more information about this error, try `rustc --explain E0277`.
```
### Desired output
```
error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> src/main.rs:2:15
  |
2 |     let _x = &str::from("value");
  |               ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `str`
  = note: the return type of a function must have a statically known size

For more information about this error, try `rustc --explain E0277`.
```
### Rationale and extra context

The first suggestion doesn't help and will just suggest stacking `&[mut]`s without solving the problem. It should be removed.

### Other cases

_No response_

### Rust Version
```
1.84.0-nightly

(2024-10-21 4392847410ddd67f6734)
```
### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by estebank**
When encountering `&str::from("value")` do not suggest `&&str::from("value")`.

Fix #132041.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs [hunk 1]
```diff
@@ -1209,6 +1209,15 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             c @ ObligationCauseCode::WhereClauseInExpr(_, _, hir_id, _)
                 if self.tcx.hir().span(*hir_id).lo() == span.lo() =>
             {
+                if let hir::Node::Expr(expr) = self.tcx.parent_hir_node(*hir_id)
+                    && let hir::ExprKind::Call(base, _) = expr.kind
+                    && let hir::ExprKind::Path(hir::QPath::TypeRelative(ty, _)) = base.kind
+                    && ty.span == span
```
Comment (Nadrieril) [inline compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:1215]: I'm not following the logic: where in this condition are you checking that the expression is already borrowed?

Comment (estebank) [reply compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:1215]: I can add the check for the existing borrow one level up, but I also wanted to avoid suggesting `&str::from("")` when we have `str::from("")` as that won't help either.

Comment (Nadrieril) [reply compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:1215]: I see, so the comment below should be saying something like "don't suggest borrowing when this wouldn't fix the problem (because the type is specified by a path instead of inferred)"?

Comment (Nadrieril) [reply compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:1215]: Ah I understand my confusion: I had understood "Do not suggest borrowing when we already do so" as "Do not suggest borrowing when the expression is already borrowed"

Comment (Nadrieril) [reply compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:1215]: Coming back to this: I don't follow what this branch is catching. Can you rephrase the comment to be more explicit about what the branch means plz?

**Comment by apiraino** [comment]
given this [comment](https://github.com/rust-lang/rust/pull/132469#discussion_r1993959985), 

@rustbot author

File: compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs]: Hi! Sorry, I had to spend time building context back for myself as well. The whole story:

When trying to find if any trait provides an method or associated function, the resolver needs to keep all trait impls and their obligations in consideration. When evaluating something like `Ty::func()`, we will inject obligations coming from `Ty` with the `HirId` and `Span` corresponding to that `PathSegment`, like `Sized` or any of its `where`-bound clauses. Because `<str as From<&str>::from` doesn't exist, we hit that unmet obligation. So far this is the normal behavior.

In `suggest_add_reference_to_arg`, we try to identify expressions where the obligation isn't met for the found type, but is for the borrow of the found type. So if you have `foo(bar)` we suggest `foo(&mut bar)` if appropriate. But that logic kind of assumed that the source of the obligation was an expression, and not a path segment of a type relative path (like `Foo::bar`, as opposed to type absolute paths like `<Foo as Trait>::bar` or `<Foo>::bar`. Because of how rare this case was in our test suite, it seemed like the only situation where the logic I initially added would only ever trigger in a situation like `&Foo::bar()`, so it didn't seem necessary to make additional checks for the borrow actually being there (which is what was quite confusing of the code as written). The reason this "fixed" the issue was because by having the early return we ensured we didn't suggest the spurious `&`.

Looking at the changes and situation with fresh eyes, I noticed a couple of cases where we could hit that new logic in item signatures as well, so I added more checks (namely, the expected one that was missing) *and* added structured suggestions, as we now had all of the spans we needed for that.

File: tests/ui/suggestions/dont-suggest-borrowing-existing-borrow.stderr
Comment (estebank) [inline tests/ui/suggestions/dont-suggest-borrowing-existing-borrow.stderr:1]: I noticed [there's another case](https://github.com/rust-lang/rust/issues/143393) that this PR doesn't account for, which provides an incorrect suggestion:

```
error[E0277]: the trait bound `S: Trait` is not satisfied
  --> $DIR/dont-suggest-borrowing-existing-borrow.rs:17:13
   |
LL |     let _ = S::foo();
   |             ^ the trait `Trait` is not implemented for `S`
   |
help: consider borrowing here
   |
LL |     let _ = &S::foo();
   |             +
LL |     let _ = &mut S::foo();
   |             ++++
```

**Review by Nadrieril — APPROVED** [review]
LGTM, r=me unless you want to also fix that other case

**Comment by estebank** [comment]
@bors r=Nadrieril

@Nadrieril I opened a ticket for that other case. I'll address it later or (even better) someone might snipe it as it is a tricky but not complex case.
